kata-agent: add network gRPC support

egernst opened the pull request:
Introduce support for setting up interface and routes when working with
Kata agent.

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst commented:
Functioning network gRPC calls dependent  on https://github.com/kata-containers/agent/pull/126

egernst commented:
This is dependent on the gRPC changes landing in kata-containers/agent and being re-vendorded.

jodh-intel commented on kata_agent.go:
```diff
@@ -295,6 +295,82 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
+		return nil
+	}
+
+	for _, endpoint := range networkNS.Endpoints {
```
Random thought: this code is pretty similar to:

- https://github.com/containers/virtcontainers/blob/master/hyperstart_agent.go#L173

I wonder if it would be worth refactoring to make a generic function in `agent.go` that accepted a `Pod` and a function to operate on the `Endpoint`'s?

egernst replied:
It is similar in that it loops over the endpoints and addresses, but with the protocols being different the implementation is different enough (operating on an interface and route at a time) that I think it may be a bit painful for a potentially short term code reuse benefit.  WDYT?

egernst commented on kata_agent.go:
```diff
@@ -295,6 +295,82 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
+		var route grpc.Route
+		var ipAddresses []*grpc.IpAddress
+		for _, addr := range endpoint.Properties().Addrs {
+			// Skip IPv6 because not supported
```
Yes: https://github.com/containers/virtcontainers/issues/579

egernst commented on kata_agent.go:
```diff
@@ -330,6 +406,10 @@ func (k *kataAgent) startPod(pod Pod) error {
 		hostname = hostname[:maxHostnameLen]
 	}
 
+	if err := k.sendNetworkInterfaceAndRouteRequests(pod); err != nil {
```
Hey @jodh-intel - I started to do this a couple of times and am having a hard time.  The current "mocking" in kata_agent_test.go is pretty limited (return nil, nil), so it is pretty hard to exercise without re-implementing a lot of logic from the existing kata agent in the test file.  I now recall all of the mock discussions that you were driving back in the CC-3.0 pre-release time frame.  

Let me think on how to handle this pragmatically.  Let me know if you have suggestion in the meantime.

jodh-intel replied:
I'm not sure I understand. We don't necessarily need 100% coverage but this PR adds a relatively big function with no tests. I don't think we should be attempting to land significant features without tests.

The approach I'd take is to break `sendNetworkInterfaceAndRouteRequests()` into smaller functions (for improved clarity, code complexity and testing) and you should then be able to unit-test most of those. But even as it is, can't you create test functions to atleast handle?:

- `fetchPodNetwork()` failing.
- `networkNS.NetNsPath == ""`.
- `addr.IP.IsLoopback() == true` (force it with `link, err := netlink.LinkByName("lo")` ?)

egernst replied:
Yes, you are right. I will refactor and add tests today.

egernst commented:
Tested with latest agent and this PR now includes revendoring for latest gRPC changes.  @sboeuf @jodh-intel  can you take a look?  Would be good to get working networking in...

sboeuf commented:
@egernst I'll review it today !

jodh-intel commented:
Yep - would be great to get networking in, once we have tests for it.

sboeuf commented on kata_agent.go:
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
+	//
+	// Setup network interfaces and routes
+	//
+	networkNS, err := pod.storage.fetchPodNetwork(pod.id)
```
Note: This should be part of the `pod` and we should be able to access those infos from in-memory. Opened an issue here: https://github.com/containers/virtcontainers/issues/628
Will be fixed in another PR

egernst replied:
ack

egernst replied:
done.

sboeuf replied:
You can update this cause you don't need to fetch the pod network anymore since #628 has been solved by #629

sboeuf commented on kata_agent.go:
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
+	interfaces, routes, err := k.generateKataInterfacesAndRoutes(networkNS)
+	if err != nil {
+		return err
+	}
```
Please add a blank line to let the code breathe :smile:
And globally, this comment applies to every place in the code where there are not enough blank lines (for readability)

egernst replied:
Will add some air.

sboeuf commented on kata_agent.go:
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
+				"interface requested": fmt.Sprintf("%+v", ifc),
+				"resulting interface": fmt.Sprintf("%+v", resultingInterface),
+			}).WithError(err).Error("update interface request failed")
+			continue
```
Why don't we fail here ? Have we agreed on ignoring failures here ?

egernst replied:
I agree on logging the error and continuing to apply remaining interfaces and routes.  I think that this is an error, but not necessarily catastrophic -- user can get into the workload and then assess what failed in setup.  WDYT?

sboeuf replied:
Well, only logging without returning this error might be confusing since the user will expect everything to work as expected if he does not check the logs. And he might not understand why he does not get what he expects inside the container.
We could have a retry mechanism (maybe retry 5 times with `100ms` between each time), so that we really make sure there is a real issue here, that would worth throwing an error.

sboeuf replied:
@sameo @jodh-intel WDYT ?

sameo replied:
@egernst @sboeuf If an interface or a route failed to be set, it's likely that networking will be at least partly dysfunctional, leaving the end-user with a potentially non working pod. By only logging this and not failing we'd give a broken pod to the upper layers of the stack and pretend it's all fine, which is bad from a end user experience perspective. In other words, we should log **and** fail here and the pod must not be passed upward. It should be stopped at least.

jodh-intel replied:
+1 for failing - if we continue, surely the system is in an unknown state (bad)? Turning it round, do we ever think this might fail? If so, why?

egernst replied:
Agreed. Thanks guys.

egernst replied:
I should have stopped with error in the first place, mea culpa.  It won't work today as I didn't handle conflicting routes (for the initial interface route in the guest).    I have a new PR, https://github.com/kata-containers/agent/pull/148, R on the agent that I started (not ready just yet) which will act on a set of routes instead, acting fully declarative.  Without this I don't think we can really test networking in a container without bypassing conflict errors (because the initial interface route already exists).  This'll simplify our handling quite a bit --   I'll try to get this ready for merge today.

sboeuf commented on kata_agent.go:
```diff
@@ -335,6 +410,45 @@ func (k *kataAgent) startPod(pod Pod) error {
+			k.Logger().WithFields(logrus.Fields{
+				"route requested": fmt.Sprintf("%+v", route),
+			}).WithError(err).Error("update route request failed")
+			continue
```
Why don't we fail here ? Have we agreed on ignoring failures here ?

sboeuf commented on kata_agent.go:
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
+			}
+			ipAddresses = append(ipAddresses, &ipAddress)
+		}
+		ifc := grpc.Interface{
```
You can do
```go
ifc := &grpc.Interface{
```
so that later in the code you can do
```go
ifaces = append(ifaces, ifc)
```

egernst replied:
yawn.  Sure. :)

egernst commented on kata_agent.go:
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
+		ifc := grpc.Interface{
+			IPAddresses: ipAddresses,
+			Device:      endpoint.Name(),
+			Name:        endpoint.Name(),
```
Done: https://github.com/kata-containers/agent/issues/147

sboeuf replied:
Thanks !

egernst commented:
@sboeuf -- updated.

egernst commented on kata_agent.go:
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
+			}
+
+			route.Gateway = r.Gw.String()
+			if route.Gateway == "<nil>" {
```
Agreed

egernst commented:
see https://github.com/kata-containers/agent/pull/148

egernst commented on kata_agent.go:
```diff
@@ -300,6 +300,81 @@ func (k *kataAgent) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
 	return prepareAndStartShim(pod, k.shim, c.id, req.ExecId, k.state.URL, cmd)
 }
 
+func (k *kataAgent) generateKataInterfacesAndRoutes(networkNS NetworkNamespace) ([]*grpc.Interface, []*grpc.Route, error) {
```
apparently I really like to push the kata brand, in this case in error.  Will update...

egernst commented:
waiting for kata-containers/agent#148 to land so I can update the revendoring.

In the meantime, addressed feedback from @sameo and @jodh-intel and @sboeuf .

egernst commented:
@sboeuf  PTAL?

egernst commented:
@jodh-intel @sboeuf  - updated based on updated pod structure.  PTAL

sboeuf commented:
LGTM

jodh-intel commented on kata_agent.go:
```diff
@@ -335,6 +412,45 @@ func (k *kataAgent) startPod(pod Pod) error {
+		resultingInterface, err := k.sendReq(ifcReq)
+		if err != nil {
+			k.Logger().WithFields(logrus.Fields{
+				"interface requested": fmt.Sprintf("%+v", ifc),
```
Nit: there is an informal understanding [*] that all log fields should be single-word (so I'd just replace spaces with hyphens here and below).

---

[*] - yep, we should make expectations formal by writing a doc! ;)

sameo commented:
LGTM

sameo commented:
@sboeuf @egernst Thanks for the PR. Let's merge it once CI passes.
